### ISSUE TITLE
Remote files support

### ISSUE THREAD

**Issue by jvmlet** [simov/markdown-viewer#8]
Hi
Does this extension support remote  files ?
For me it works only when I open the local md file  with Chrome, remote  file (http://host/file.md) is rendered as is.
Thanks

**Comment by simov** [comment]
Yep, here is an example of a markdown file from this repository: https://raw.githubusercontent.com/simov/markdown-viewer/master/syntax.md

**Comment by simov** [comment]
And here is how it looks like rendered with this extension:
![markdown](https://cloud.githubusercontent.com/assets/1694112/19218829/9dfded62-8e0c-11e6-85f4-f38f7f6b2334.png)

**Comment by jvmlet** [comment]
Ok, looks like my particular URL is tricky  :

URL :  **http://host:port/path/file.md/\*view\*/**

Response headers:

Accept-Ranges:bytes
**Content-Disposition:inline; filename=file.md**
Content-Encoding:gzip
Content-Length:2386
Content-Type:text/plain

Any chance to support this ?
Thanks

**Comment by simov** [comment]
Can you provide an actual link to test with, or at least some test location where I can reproduce it?

**Comment by jvmlet** [comment]
Unfortunately, no, 
It's on private network, but to simulate this, you can have REST controller that `generates` the  md file and sends it with  `Content-Disposition` response header.

**Comment by jvmlet** [comment]
BTW, I don't see integration tests in your repository, if you have it, I would be able to PR the controllers' code (preferably with Spring MVC )

**Comment by simov** [comment]
Currently the extension matches only on:

```js
if (/.*\/.*\.(?:markdown|mdown|mkdn|md|mkd|mdwn|mdtxt|mdtext|text)(?:#.*)?$/.test(tab.url))
```

That's the first step, and making it match on `.md` anywhere in the URL will be unreliable. 

Only after that the extension can try to make xhr request to that URL and check the headers for `Content-Disposition:inline; filename=file.md`

So again the first and foremost problem is to reliably determine if the URL should be treated as a markdown file. To get the headers upfront would require completely different set of Chrome APIs and implementation.

**Comment by jvmlet** [comment]
I totally  agree that matching on md extension anywhere in the URL  is not the option...
Another thought is to have `enforce` button to trigger the viewer...

**Comment by jvmlet** [comment]
More sophisticated solution would be to expose the URL matching regex expression**_s_** as configuration option, providing your hard-coded

 ````javascript
/.*\/.*\.(?:markdown|mdown|mkdn|md|mkd|mdwn|mdtxt|mdtext|text)(?:#.*)?$/
````

 as default.

**Comment by simov** [comment]
> Another thought is to have enforce button to trigger the viewer...

That's not a bad idea. However this will require the extension to be implemented as [browserAction](https://developer.chrome.com/extensions/browserAction) instead of [pageAction](https://developer.chrome.com/extensions/pageAction) like it is right now. Because right now if the URL is not matched the extension is simply inactive.

But given that the Chrome team abandoned the pageAction's like they used to work, that might not be a huge change here.

Exposing the regex as an option is not a bad idea either, and I'm thinking about implementing an options page at some point.

I'll definitely take into consideration these two options, but I can't give you any estimate on when this is going to be implemented because I want to migrate to Mithril 1.x and implement Material design in the popup first.

**Comment by simov** [comment]
Actually @jvmlet if you want to help me with this, can you create a small HTTP server that reproduces the behavior you have problems with?

A gist would be great - you can add multiple files in it if needed, and I can download and run the whole thing easily.

**Comment by jvmlet** [comment]
Let me check If I can host the sample online ...

**Comment by jvmlet** [comment]
The sample REST service is here : https://md-rest-download.appspot.com/md

**Comment by simov** [comment]
Thanks!

I got it working but I think that the best way to implement this as a feature is to have an options page where the user can optionally add a matching pattern or something like that.

I'll have to think about it, and probably test a few other ideas as well.

Thanks for the feedback! I'll let you know as soon as I have something.

**Comment by jvmlet** [comment]
Great, thanks
BTW, Markdown Preview Plus solves this issue  by making addition ajax request, not sure how it decides when to make it ...

**Comment by simov** [comment]
Yes, that's a good way to implement it. 

I am considering a major change to the permissions required by the extension, and this will affect the way it works. Right now it says: _Read and change all your data on the websites you visit_ which might turn away some people.

**Comment by simov** [comment]
@jvmlet I just published v2.3 this is a really huge update to the extension, but in a nutshell the extension no longer requires any additional permissions for rendering local file URLs.

Any additional remote origin should be enabled manually, here is an example of the new options page:

![origins](https://cloud.githubusercontent.com/assets/1694112/19836483/735b51d6-9eaa-11e6-871b-ba553f966371.png)

This page also allows you to set a matching regular expression for the path. Your example server works with the above configuration.

I'll be happy to hear your feedback on this.

**Comment by jvmlet** [comment]
Looks great , thanks.
One note about the  origin: it would be nice to have protocol part as optional, or regex.
At most cases, user will have to specify two origins for http and https with the same pattern.
Thanks again for the great extension.

**Comment by simov** [comment]
Hey @jvmlet I just published version 2.6 where you can use `*` to add origin for both `https` and `http`. Take a look at the [updated docs](https://github.com/simov/markdown-viewer#add-origin).
